src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.SyntaxAndDeclarationManager.cs

```diff
@@ -0,0 +1,587 @@
+                    var path = (string)fileToken.Value;
+                    if (path == null)
+                    {
+                        // If there is no path, the parser should have some Diagnostics to report.
+                        Debug.Assert(tree.GetDiagnostics().Any(d => d.Severity == DiagnosticSeverity.Error));
+                        continue;
+                    }
+
+                    var resolvedFilePath = resolver.ResolveReference(path, baseFilePath: null);
+                    if (resolvedFilePath == null)
+                    {
+                        diagnostics.Add(
+                            messageProvider.CreateDiagnostic(
+                                (int)ErrorCode.ERR_NoSourceFile,
+                                fileToken.GetLocation(),
+                                path,
+                                CSharpResources.CouldNotFindFile));
+                    }
+                    else if (!loadedSyntaxTreeMapBuilder.ContainsKey(resolvedFilePath))
+                    {
+                        var readTextDiagnostics = new List<Diagnostic>();
+                        var code = resolver.ReadText(resolvedFilePath, fileToken.GetLocation(), readTextDiagnostics);
+                        if (code == null)
+                        {
+                            Debug.Assert(readTextDiagnostics.Any());
+                            diagnostics.AddRange(readTextDiagnostics);
+                        }
+                        else
+                        {
+                            loadedTree = SyntaxFactory.ParseSyntaxTree(
+                                code,
+                                tree.Options, // Use ParseOptions propagated from "external" tree.
+                                resolvedFilePath);
+
+                            // All #load'ed trees should have unique path information.
+                            loadedSyntaxTreeMapBuilder.Add(loadedTree.FilePath, loadedTree);
+
+                            AppendAllSyntaxTrees(
+                                treesBuilder,
+                                loadedTree,
+                                options,
+                                messageProvider,
+                                isSubmission,
+                                ordinalMapBuilder,
+                                loadDirectiveMapBuilder,
+                                loadedSyntaxTreeMapBuilder,
+                                declMapBuilder,
+                                ref declTable,
+                                ref reuseReferenceManager);
+                        }
+                    }
+                    else
+                    {
+                        // The path resolved, but we've seen this file before,
+                        // so don't attempt to load it again.
+                        Debug.Assert(diagnostics.IsEmptyWithoutResolution);
+                        diagnostics.Free();
+                        continue;
```

Comment:
Shouldn't the file reference be included in loadDirectives, even though the file has already been loaded elsewhere in the graph?